lxd/instance_state.go

```diff
@@ -137,7 +138,12 @@ func doInstanceStatePut(inst instance.Instance, req api.InstanceStatePut) error
 			timeout = 0
 		}
 
-		return inst.Restart(time.Duration(timeout))
+		err := createSnapshotIfScheduled(inst, d, timeout)
+		if err == nil {
+			return inst.Restart(time.Duration(timeout))
+		} else {
+			return fmt.Errorf("Creating scheduled @reboot snapshot failed with message '%s', aborting reboot", err)
+		}
```

Comment:
Sorry I missed that part in my earlier review.

Instead of having a `@reboot` which we'd need to plug in whenever we restart an issue (either here in PUT or when the instance self-reboots using `reboot` inside the instance), I think it'd be best to drop `@reboot` and instead introduce `@startup` which will be triggering a snapshot every time the instance starts or restarts.

The best place to handle that would be the `startCommon` function of the LXC driver and the `Start` function of the QEMU driver. Basically just calling for a snapshot to be created just before starting up the instance if `snapshots.schedule` is set to `@startup`.

Again, sorry for not having caught this earlier :(